src/transformers/models/rag/modeling_tf_rag.py

```diff
@@ -0,0 +1,2316 @@
+
+    def set_retriever(self, retriever: RagRetriever):
+        self.rag.retriever = retriever
+
+    def adjust_logits_during_generation(self, logits, cur_len, max_length):
+        return self.rag.generator.adjust_logits_during_generation(logits, cur_len=cur_len, max_length=max_length)
+
+    # Adapted from https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_tf_bart.py
+    def prepare_inputs_for_generation(
+        self, decoder_input_ids, past, attention_mask, use_cache, encoder_outputs, doc_scores, n_docs=None, **kwargs
+    ) -> Dict:
+        assert past is not None and len(past) in {1, 2}, f"past has to be an iterable of length 1,2 got {past}"
+
+        if len(past) == 1:
+            assert isinstance(past[0], tf.Tensor)
+            encoder_outputs = TFBaseModelOutput(last_hidden_state=past[0])
+            decoder_cached_states = None
+        else:
+            assert len(past) == 2
+            # Note: encoder_outputs is never changed by Bart as a generator
+            encoder_outputs, decoder_cached_states = past
+
+            if isinstance(encoder_outputs, tuple):
+                assert isinstance(encoder_outputs[0], tf.Tensor)
+                encoder_outputs = TFBaseModelOutput(last_hidden_state=encoder_outputs[0])
+            elif isinstance(encoder_outputs, tf.Tensor):
+                encoder_outputs = TFBaseModelOutput(last_hidden_state=encoder_outputs)
+
+            assert (
+                decoder_cached_states
+            ), f"decoder cached states must be truthy. got {decoder_cached_states} from the 2nd element of past"
+            # if past is defined cut decoder_input_ids to last token
+            decoder_input_ids = decoder_input_ids[:, -1:]
+
+        assert isinstance(
+            encoder_outputs, TFBaseModelOutput
+        ), f"encoder_outputs should be a TFBaseModelOutput, Instead got {type(encoder_outputs)}."
+        return {
+            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
+            "encoder_outputs": encoder_outputs,
+            "doc_scores": doc_scores,
+            "context_attention_mask": attention_mask,
+            "decoder_input_ids": decoder_input_ids,
+            "past_key_values": decoder_cached_states,  # This is due to TFBart and is the main difference to Pytorch's RAG
+            "use_cache": use_cache,  # change this to avoid caching (presumably for debugging)
+            "do_marginalize": True,
+            "n_docs": n_docs,
+        }
+
+    @property
+    def retriever(self):
+        return self.rag.retriever
+
+    @property
+    def generator(self):
+        return self.rag.generator
+
+    @property
+    def question_encoder(self):
+        return self.rag.question_encoder
+
+    @staticmethod
+    def _reorder_cache(past, beam_idx):
+        """Reorders cache for generation. BART-inspired but we need to take care of the extra dimension for docs"""
+
+        def tf_index_select(input_, dim, indices):
```

Comment:
I'm fine with this -> could probably be sped-up by a tensor-only operation, but ok for now IMO.